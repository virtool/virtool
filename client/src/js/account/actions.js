/**
 * Redux action creators for use with the active user account.
 *
 * @module account/actions
 */
import { simpleActionCreator } from "../utils";
import {
    GET_ACCOUNT,
    UPDATE_ACCOUNT,
    UPDATE_ACCOUNT_SETTINGS,
    CHANGE_ACCOUNT_PASSWORD,
    GET_API_KEYS,
    CREATE_API_KEY,
    UPDATE_API_KEY,
    REMOVE_API_KEY,
    LOGOUT,
    CLEAR_API_KEY
} from "../actionTypes";

/**
 * Returns action that can trigger an API call for getting the current account data.
 *
 * @func
 * @returns {object}
 */
export const getAccount = simpleActionCreator(GET_ACCOUNT.REQUESTED);

/**
 * Returns an action that can trigger an API call for updating the current account.
 *
 * @func
 * @param update {object} the update to apply to the account
 * @returns {object}
 */
export const updateAccount = (update) => ({
    type: UPDATE_ACCOUNT.REQUESTED,
    update
});

/**
 * Returns an action that can trigger an API call for updating the settings for the current account.
 *
 * @func
 * @param update {object} the update to apply to the account
 * @returns {object}
 */
export const updateAccountSettings = (update) => ({
    type: UPDATE_ACCOUNT_SETTINGS.REQUESTED,
    update
});

/**
 * Returns an action that can trigger an API call for changing the password for the current account.
 *
 * @func
 * @param oldPassword {string} the old account password (used for verification)
 * @param newPassword {string} the new account password
 * @returns {object}
 */
export const changePassword = (oldPassword, newPassword) => ({
    type: CHANGE_ACCOUNT_PASSWORD.REQUESTED,
    oldPassword,
    newPassword
});

/**
 * Returns action that can trigger an API call for getting the API keys owned by the current account.
 *
 * @func
 * @returns {object}
 */
export const getAPIKeys = simpleActionCreator(GET_API_KEYS.REQUESTED);

/**
 * Returns action that can trigger an API call for creating a new API key for the current account.
 *
 * @func
 * @param name {string} a name for the API key
 * @param permissions {object} permissions configuration object for the new API key
 * @returns {object}
 */
export const createAPIKey = (name, permissions) => ({
    type: CREATE_API_KEY.REQUESTED,
    name,
    permissions
});

/**
 * Clears temporarily stored new API keys.
 *
 * @func
 * @returns {object}
 */
export const clearAPIKey = simpleActionCreator(CLEAR_API_KEY);

/**
 * Returns action that can trigger an API call for updating the permissions for an API key owned by the current account.
 *
 * @func
 * @param keyId {string} the unique id for the API key
 * @param permissions {object} permissions configuration object for the new API key
 * @returns {object}
 */
export const updateAPIKey = (keyId, permissions) => ({
    type: UPDATE_API_KEY.REQUESTED,
    keyId,
    permissions
});

/**
 * Returns action that can trigger an API call for removing an API key owned by the current account.
 *
 * @func
 * @param keyId {string} the unique id for the API key
 * @returns {object}
 */
export const removeAPIKey = (keyId) => ({
    type: REMOVE_API_KEY.REQUESTED,
    keyId
});

/**
 * Returns action that can trigger an API call that will logout the current session.
 *
 * @func
 * @returns {object}
 */
export const logout = simpleActionCreator(LOGOUT.REQUESTED);
