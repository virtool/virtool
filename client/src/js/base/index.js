import { AlgorithmSelect } from "./AlgorithmSelect";
import { AutoProgressBar, ProgressBar } from "./ProgressBar";
import { bsStyles } from "./utils";
import { Checkbox } from "./Checkbox";
import { Flex, FlexItem } from "./Flex";
import { Help } from "./Help";
import { Icon } from "./Icon";
import { Identicon } from "./Identicon";
import { Input } from "./Input";
import { InputError } from "./InputError";
import { InputSave } from "./InputSave";
import { ListGroupItem } from "./ListGroupItem";
import { NotFound } from "./NotFound";
import { Panel } from "./Panel";
import { Radio } from "./Radio";
import { RelativeTime } from "./RelativeTime";
import { RemoveBanner } from "./RemoveBanner";
import { RemoveModal } from "./RemoveModal";
import { SaveButton } from "./SaveButton";
import { ScrollList } from "./ScrollList";
import { Tooltip } from "./Tooltip";
import { UploadBar } from "./UploadBar";
import { ViewHeader } from "./ViewHeader";

export * from "./Alert";
export * from "./Attribution";
export * from "./Badge";
export * from "./Box";
export * from "./Breadcrumb";
export * from "./Breakpoints";
export * from "./Button";
export * from "./ButtonGroup";
export * from "./ButtonToolbar";
export * from "./Code";
export * from "./Container";
export * from "./ExternalLink";
export * from "./Label";
export * from "./Loader";
export * from "./LoadingPlaceholder";
export * from "./Logo";
export * from "./Modal";
export * from "./NoneFound";
export * from "./Table";
export * from "./SectionHeader";
export * from "./SlashList";
export * from "./TabLink";
export * from "./Tabs";
export * from "./Toolbar";
export * from "./ViewHeader";
export * from "./SelectBox";

export {
    AlgorithmSelect,
    AutoProgressBar,
    bsStyles,
    Checkbox,
    Flex,
    FlexItem,
    Help,
    Icon,
    Identicon,
    Input,
    InputError,
    InputSave,
    ListGroupItem,
    NotFound,
    Panel,
    ProgressBar,
    Radio,
    RelativeTime,
    RemoveBanner,
    RemoveModal,
    SaveButton,
    ScrollList,
    Tooltip,
    UploadBar,
    ViewHeader
};
