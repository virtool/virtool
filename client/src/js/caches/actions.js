import { GET_CACHE } from "../app/actionTypes";

export const getCache = cacheId => ({
    type: GET_CACHE.REQUESTED,
    cacheId
});
