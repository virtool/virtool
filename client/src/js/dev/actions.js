import { POST_DEV_COMMAND } from "../app/actionTypes";

export const postDevCommand = command => ({ type: POST_DEV_COMMAND.REQUESTED, command });
