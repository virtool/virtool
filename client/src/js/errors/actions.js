/**
 * Redux action creator for clearing errors.
 *
 * @module error/actions
 */
import { CLEAR_ERROR } from "../actionTypes";

/**
 * Returns action that clears specific error in the store.
 *
 * @func
 * @returns {object}
 */
export const clearError = (error) => ({
    type: CLEAR_ERROR,
    error
});
