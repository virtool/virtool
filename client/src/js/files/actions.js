/**
 * Redux action creators for use with file management.
 *
 * @module files/actions
 */

import { simpleActionCreator } from "../utils";
import {
    WS_UPDATE_FILE,
    WS_REMOVE_FILE,
    FIND_FILES,
    REMOVE_FILE,
    UPLOAD,
    UPLOAD_PROGRESS,
    HIDE_UPLOAD_OVERLAY
} from "../actionTypes";

/**
 * Returns an action that should be dispatched when a file document is updated via websocket.
 *
 * @func
 * @param data {object} the data passed in the websocket message
 * @returns {object}
 */
export const wsUpdateFile = (data) => ({
    type: WS_UPDATE_FILE,
    data
});

/**
 * Returns an action that should be dispatched when a file document is removed via websocket.
 *
 * @func
 * @param data {object} the data passed in the websocket message
 * @returns {object}
 */
export const wsRemoveFile = (data) => ({
    type: WS_REMOVE_FILE,
    data
});

/**
 * Returns an action that can trigger an API call that finds file documents.
 *
 * @func
 * @param fileType {string} a file type to filter the returned documents by
 * @param page {number} which page of results to return
 * @returns {object}
 */
export const findFiles = (fileType, page) => ({
    type: FIND_FILES.REQUESTED,
    fileType,
    page
});

/**
 * Returns an action that can trigger the upload of a file to the server.
 *
 * @func
 * @param localId {string} the local id for the upload (NOT the fileId)
 * @param file {object} file to be uploaded
 * @param fileType {string} file type
 * @param onProgress {function} callback function for uploadProgress API call
 * @returns {object}
 */
export const upload = (localId, file, fileType, onProgress) => ({
    type: UPLOAD.REQUESTED,
    localId,
    file,
    fileType,
    onProgress
});

/**
 * Returns an action that can trigger an API call that removes a file by its fileId.
 *
 * @func
 * @param fileId {string} the unique id for the file
 * @returns {object}
 */
export const removeFile = (fileId) => ({
    type: REMOVE_FILE.REQUESTED,
    fileId
});

/**
 * Returns and action that updates the progress of an ongoing upload.
 *
 * @func
 * @param localId {string} the local id for the upload (NOT the fileId)
 * @param progress {number} the new progress value
 * @returns {object}
 */
export const uploadProgress = (localId, progress) => ({
    type: UPLOAD_PROGRESS,
    localId,
    progress
});

/**
 * A simple action that causes the upload dialog overlay to be hidden.
 *
 * @func
 * @returns {object}
 */
export const hideUploadOverlay = simpleActionCreator(HIDE_UPLOAD_OVERLAY);
