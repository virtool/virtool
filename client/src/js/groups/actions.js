import { simpleActionCreator } from "../utils";
import { LIST_GROUPS, CREATE_GROUP, SET_GROUP_PERMISSION, REMOVE_GROUP } from "../actionTypes";

/**
 * Returns an action that triggers a request for all groups from the API.
 *
 * @func
 */
export const listGroups = simpleActionCreator(LIST_GROUPS.REQUESTED);

/**
 * Returns an action that triggers a API request to create a group with the given `groupId`.
 *
 * @func
 * @param groupId {string} the id for the new group
 * @returns {object}
 */
export const createGroup = (groupId) => ({
    type: CREATE_GROUP.REQUESTED,
    groupId
});

/**
 * Returns action that can trigger an API request for modifying group permissions.
 *
 * @param groupId {string} the id for the specific group
 * @param permission {string} the specific permission field
 * @param value {boolean} is checked or not
 * @returns {object}
 */
export const setGroupPermission = (groupId, permission, value) => ({
    type: SET_GROUP_PERMISSION.REQUESTED,
    groupId,
    permission,
    value
});

/**
 * Returns an action that triggers a API request to remove a group with the given `groupId`.
 *
 * @func
 * @param groupId {string} the id for the new group
 * @returns {object}
 */
export const removeGroup = (groupId) => ({
    type: REMOVE_GROUP.REQUESTED,
    groupId
});
