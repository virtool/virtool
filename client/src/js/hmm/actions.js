import { simpleActionCreator } from "../utils";
import { GET_HMM, INSTALL_HMMS, FETCH_HMMS } from "../actionTypes";

export const fetchHmms = simpleActionCreator(FETCH_HMMS);

export const getHmm = (hmmId) => ({
    type: GET_HMM.REQUESTED,
    hmmId
});

export const installHMMs = simpleActionCreator(INSTALL_HMMS.REQUESTED);
