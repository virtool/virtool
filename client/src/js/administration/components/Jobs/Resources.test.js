import ResourcesContainer, {
    getErrorMessage,
    LimitLabel,
    Resources
} from "./Resources";
import { InputError, Alert } from "../../../base";
import * as actions from "../../actions";
import * as jobActions from "../../../jobs/actions";
import * as errorActions from "../../../errors/actions";

describe("<Resources />", () => {
    const initialState = {
        jobs: {
            resources: {
                mem: { total: 1000, available: 500 },
                proc: [10, 15, 20]
            }
        },
        settings: {
            data: {
                create_sample_inst: 10,
                create_sample_mem: 5,
                create_sample_proc: 3,
                build_index_inst: 10,
                build_index_mem: 5,
                build_index_proc: 3,
                create_subtraction_inst: 10,
                create_subtraction_mem: 5,
                create_subtraction_proc: 3,
                pathoscope_bowtie_inst: 10,
                pathoscope_bowtie_mem: 5,
                pathoscope_bowtie_proc: 3,
                nuvs_inst: 10,
                nuvs_mem: 5,
                nuvs_proc: 3,
                proc: 3,
                mem: 5
            }
        },
        error: null
    };
    const store = mockStore(initialState);
    let props;
    let wrapper;

    it("renders correctly", () => {
        wrapper = shallow(<ResourcesContainer store={store} />).dive();
        expect(wrapper).toMatchSnapshot();
    });

    it("getErrorMessage(): returns error message when [isError=true], otherwise returns false", () => {
        let result = getErrorMessage(true, 1, 10);
        const expected = "Value must be between 1 and 10";
        expect(result).toEqual(expected);

        result = getErrorMessage(false, 1, 10);
        expect(result).toEqual(null);
    });

    it("renders LimitLabel subcomponent", () => {
        props = {
            label: "test",
            available: 5,
            unit: "GB"
        };
        wrapper = shallow(<LimitLabel {...props} />);
        expect(wrapper).toMatchSnapshot();
    });

    it("getDerivedStateFromProps", () => {
        let nextProps = { mem: 3, proc: 2 };
        let prevState = { mem: 1, proc: 2 };
        let result = Resources.getDerivedStateFromProps(nextProps, prevState);
        let expected = { errorMem: false };
        expect(result).toEqual(expected);

        nextProps = { mem: 1, proc: 4 };
        prevState = { mem: 1, proc: 2 };
        result = Resources.getDerivedStateFromProps(nextProps, prevState);
        expected = { errorProc: false };
        expect(result).toEqual(expected);

        nextProps = { mem: 1, proc: 2 };
        prevState = { mem: 1, proc: 2 };
        result = Resources.getDerivedStateFromProps(nextProps, prevState);
        expect(result).toEqual(null);
    });

    describe("componentDidUpdate", () => {

        it("Update entry values whose server default values exceed resource maximums", () => {
            props = {
                onGet: jest.fn(),
                onUpdate: sinon.spy(),
                maxMem: 10,
                maxProc: 10
            };
            wrapper = shallow(<Resources {...props} />);

            wrapper.setProps({ overResourceMax: { test_proc: 100, test_mem: 100 } });
            expect(props.onUpdate.calledWith({ name: "test_proc", value: 10 })).toBe(true);
        });

        it("Update entry values whose values exceed resource maximums", () => {
            props = {
                onGet: jest.fn(),
                onUpdate: sinon.spy(),
                overResourceMax: { test_mem: 100 },
                maxMem: 10,
                maxProc: 10,
                proc: 1,
                mem: 30
            };
            wrapper = shallow(<Resources {...props} />);
            wrapper.setProps({ overResourceMax: {} });
            expect(props.onUpdate.calledWith({ name: "mem", value: 10 })).toBe(true);

            wrapper.setProps({ overResourceMax: { foo: "bar" } });
            wrapper.setState({ mem: 1, proc: 30 });

            wrapper.setProps({ overResourceMax: {} });
            expect(props.onUpdate.calledWith({ name: "proc", value: 10 })).toBe(true);
        });

    });

    it("renders Resources component and calls onGet prop on mount", () => {
        props = {
            onGet: jest.fn(),
            proc: 3,
            mem: 5,
            minProc: 1,
            mimMem: 1,
            maxProc: 8,
            maxMem: 10,
            resources: {
                mem: { total: 1000, available: 500 },
                proc: [10, 15, 20]
            },
            error: "",
            overResourceMax: {}
        };
        wrapper = shallow(<Resources {...props} />);

        expect(wrapper).toMatchSnapshot();
        expect(props.onGet).toHaveBeenCalled();
    });

    it("renders a <LoadingPlaceholder /> component when resources data are not yet fetched", () => {
        wrapper = shallow(<Resources onGet={jest.fn()} resources={null} />);
        expect(wrapper).toMatchSnapshot();
    });

    it("renders an alert if the error props is set", () => {
        wrapper = shallow(<Resources onGet={jest.fn()} error="test-error" />);
        expect(wrapper).toMatchSnapshot();
        expect(wrapper.find(Alert).exists()).toBe(true);
    });

    it("calls setError when input is invalid", () => {
        const mockEvent = {
            preventDefault: jest.fn(),
            target: { name: "test", value: 0 }
        };

        wrapper = mount(<Resources onGet={jest.fn()} />);
        wrapper.find(InputError).at(0).prop("onInvalid")(mockEvent);

        expect(wrapper.state("errorTest")).toBe(true);
    });

    describe("dispatch functions", () => {
        let spy;
        let mockEvent;

        afterEach(() => {
            spy.restore();
        });

        it("calls updateSetting() on input submit", () => {
            spy = sinon.spy(actions, "updateSetting");
            expect(spy.called).toBe(false);

            wrapper = mount(<ResourcesContainer store={store} />);

            mockEvent = {
                name: "test",
                value: 5,
                max: 10,
                min: 1
            };

            wrapper.find(InputError).at(0).prop("onSave")({...mockEvent, value: 100});
            expect(spy.called).toBe(false);

            wrapper.find(InputError).at(0).prop("onSave")(mockEvent);
            expect(spy.calledWith("test", 5)).toBe(true);
        });

        it("calls getResources() on componentDidMount", () => {
            spy = sinon.spy(jobActions, "getResources");
            expect(spy.called).toBe(false);

            wrapper = mount(<ResourcesContainer store={store} />);

            expect(spy.calledOnce).toBe(true);
        });

        it("calls clearError() to clear error in state and in store on input change", () => {
            spy = sinon.spy(errorActions, "clearError");
            expect(spy.called).toBe(false);

            wrapper = mount(<ResourcesContainer store={store} />);

            mockEvent = {
                preventDefault: jest.fn(),
                target: { name: "test" }
            };
            wrapper.find(InputError).at(0).prop("onChange")(mockEvent);

            expect(spy.calledWith("UPDATE_SETTINGS_ERROR")).toBe(true);
        });

    });

});
