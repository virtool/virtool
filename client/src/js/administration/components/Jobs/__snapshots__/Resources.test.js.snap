// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Resources /> renders LimitLabel subcomponent 1`] = `
<h5>
  <Flex
    alignContent="stretch"
    alignItems="center"
    direction="row"
    justifyContent="flex-start"
    wrap="nowrap"
  >
    <FlexItem
      alignSelf={null}
      basis="auto"
      grow={1}
      shrink={0}
    >
      <strong>
        test
      </strong>
    </FlexItem>
    <FlexItem
      alignSelf={null}
      basis="auto"
      grow={0}
      shrink={0}
    >
      <small
        className="text-info"
      >
        5
         
        GB
         available
      </small>
    </FlexItem>
  </Flex>
</h5>
`;

exports[`<Resources /> renders Resources component and calls onGet prop on mount 1`] = `
<div>
  <AdministrationSection
    content={
      <PanelBody
        bsClass="panel"
        collapsible={false}
      >
        <LimitLabel
          available={8}
          label="CPU Limit"
          unit="cores"
        />
        <InputError
          error={null}
          initialValue={3}
          max={8}
          min={1}
          name="proc"
          noMargin={true}
          onChange={[Function]}
          onInvalid={[Function]}
          onSave={[Function]}
          type="number"
          withButton={true}
        />
        <LimitLabel
          available={10}
          label="Memory Limit (GB)"
          unit="GB"
        />
        <InputError
          error={null}
          initialValue={5}
          max={10}
          name="mem"
          noMargin={true}
          onChange={[Function]}
          onInvalid={[Function]}
          onSave={[Function]}
          type="number"
          withButton={true}
        />
      </PanelBody>
    }
    description="Set limits on the computing resources Virtool can use on the host server."
    title="Resource Limits"
  />
</div>
`;

exports[`<Resources /> renders a <LoadingPlaceholder /> component when resources data are not yet fetched 1`] = `<LoadingPlaceholder />`;

exports[`<Resources /> renders an alert if the error props is set 1`] = `
<div>
  <Alert
    bsStyle="danger"
    icon="exclamation-circle"
  >
    Resource Limit values cannot be lower than corresponding Task-specific Limits.
  </Alert>
  <AdministrationSection
    content={
      <PanelBody
        bsClass="panel"
        collapsible={false}
      >
        <LimitLabel
          label="CPU Limit"
          unit="cores"
        />
        <InputError
          error={null}
          name="proc"
          noMargin={true}
          onChange={[Function]}
          onInvalid={[Function]}
          onSave={[Function]}
          type="number"
          withButton={true}
        />
        <LimitLabel
          label="Memory Limit (GB)"
          unit="GB"
        />
        <InputError
          error={null}
          name="mem"
          noMargin={true}
          onChange={[Function]}
          onInvalid={[Function]}
          onSave={[Function]}
          type="number"
          withButton={true}
        />
      </PanelBody>
    }
    description="Set limits on the computing resources Virtool can use on the host server."
    title="Resource Limits"
  />
</div>
`;

exports[`<Resources /> renders correctly 1`] = `
<div>
  <AdministrationSection
    content={
      <PanelBody
        bsClass="panel"
        collapsible={false}
      >
        <LimitLabel
          available={3}
          label="CPU Limit"
          unit="cores"
        />
        <InputError
          error={null}
          initialValue={3}
          max={3}
          min={3}
          name="proc"
          noMargin={true}
          onChange={[Function]}
          onInvalid={[Function]}
          onSave={[Function]}
          type="number"
          withButton={true}
        />
        <LimitLabel
          available={0}
          label="Memory Limit (GB)"
          unit="GB"
        />
        <InputError
          error={null}
          initialValue={5}
          max={0}
          min={5}
          name="mem"
          noMargin={true}
          onChange={[Function]}
          onInvalid={[Function]}
          onSave={[Function]}
          type="number"
          withButton={true}
        />
      </PanelBody>
    }
    description="Set limits on the computing resources Virtool can use on the host server."
    title="Resource Limits"
  />
</div>
`;
